### PR TITLE
let azure core provide json interface

### PR DESCRIPTION
**Pull request by vhvb1989**
Making Azure Core depend on nlohmann JSON API.
Azure Core will try to find the library, if it is not found, it will be fetched as dependency with CMake.

The library will be exposed as an interface for anyone using Azure Core to be able to include json header.

This enables Json functionality for any Service

fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/1005

### CODE REVIEW

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:19]: We need to think carefully before taking a dependency and publicly exposing a third party header like this.

File: cmake-modules/add_nlohmann_json.cmake
Comment (Jinming-Hu) [inline cmake-modules/add_nlohmann_json.cmake:30]: newline

File: cmake-modules/add_nlohmann_json.cmake
Comment (Jinming-Hu) [inline cmake-modules/add_nlohmann_json.cmake:20]: Why do we download rather than install via vcpkg like other dependencies?

Comment (vhvb1989) [reply cmake-modules/add_nlohmann_json.cmake:20]: well, one of the benefits of this json lib is that it is just a header file with no src. All we need is to download the json header and included as an interface.

The end user can still get it with vcpkg or install it in the system, in that case, we would find it with `find_package` and won't download anything.

What I didn't like is to block the compilation for a missing lib that is only a header file that we can easily get. There is no multiple compilation flavors for it (like we have for other libs like libcurl) where the end user might want to get and compile the dependency with some specific config.
So, trying to simplify the customer experience.

And one last thing, is that this is still part of a temporal functionality that is not yet certified and accepted for Azure Core to GA, so, we will most likely change it

File: sdk/core/azure-core/CMakeLists.txt
Comment (ahsonkhan) [inline sdk/core/azure-core/CMakeLists.txt:65]: Can you help me understand why we are picking `interface` here? 

I am reading this to see when to use when we would use that: https://cmake.org/cmake/help/v3.0/command/target_link_libraries.html

Is this the reason for using interface here?
> If your header files but not your source files include the headers of another library, then it is an INTERFACE dependency.

Comment (Jinming-Hu) [reply sdk/core/azure-core/CMakeLists.txt:65]: I don't think it should be `INTERFACE`

Comment (ahsonkhan) [reply sdk/core/azure-core/CMakeLists.txt:65]: > We can leave out this keyword.

What are the semantics if a keyword is not provided?

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [inline cmake-modules/add_nlohmann_json.cmake:6]: Update comment. since it doesn't match what this cmake module is doing.

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [inline cmake-modules/add_nlohmann_json.cmake:11]: Good comment!

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [inline cmake-modules/add_nlohmann_json.cmake:14]: What happens here, if the application already uses a `nlohmnn_json` but with a different version (higher or lower)? Does `find_package` still find it?

I am not sure about this so this is an open question, but should we use `EXACT` match?

https://cmake.org/cmake/help/v3.0/command/find_package.html
> When the [version] argument is given Config mode will only find a version of the package that claims compatibility with the requested version (format is major[.minor[.patch[.tweak]]]). If the EXACT option is given only a version of the package claiming an exact match of the requested version may be found.

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [inline cmake-modules/add_nlohmann_json.cmake:19]:
```suggestion
    # release code only. This save us from getting the entire nlohmann source code.
```

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [inline cmake-modules/add_nlohmann_json.cmake:27]: nit: add indentation in the if-block
```suggestion
      FetchContent_Populate(json)
      add_subdirectory(${json_SOURCE_DIR} ${json_BINARY_DIR} EXCLUDE_FROM_ALL)
```

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [inline cmake-modules/add_nlohmann_json.cmake:25]: nit: consistent spacing
```suggestion
    if (NOT json_POPULATED)
```

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [inline cmake-modules/add_nlohmann_json.cmake:11]:
```suggestion
# When it is not found it, Azure Core will fetch it.
```

File: cmake-modules/add_nlohmann_json.cmake
Comment (vhvb1989) [reply cmake-modules/add_nlohmann_json.cmake:14]: Valid question. If the version is lower, CMake would consider it as not found, so it would fetch the version we need.
Then, it might collapse with customer's version, and fail on compile time.

If the version is higher, CMake consider it as FOUND and use that higher version.

So, the only problem here is when the version is lower.
I tried to use an old version (3.2) but Storage was already using 3.8 and got an error because `contains` API was missing, so I updated it.

This is one of the things we want to define for the final approach.

File: sdk/core/azure-core/CMakeLists.txt
Comment (vhvb1989) [reply sdk/core/azure-core/CMakeLists.txt:65]: sure, INTERFACE means that the target/lib is not linked and become part of Azure Core. It is only set as to be linked together with anyone linking azure core.
So, when an end-user or a service links azure-core, CMake will translate to "link azure core and also link this json lib".
The difference with PUBLIC is that the library bytes are copied directly to azure-core. Then when azure core is linked, there is no extra libs linked (behind the scenes).

The benefit is that Azure Core will not grow in size by adding the entire Json lib (as with PUBLIC)

Usually services, or anyone creating a new library on top of azure core would use PRIVATE linking, meaning that the lib will only take the required bytes from the linked library without re-exposing its functionality. Meaning that if the new library is only using one method from JSON, only that method will be copied to the new library.

If a new library would like to expose the API from an inner library, (like if Storage would like to offer JSON api), they would link it as INTERFACE to avoid copying all inside core lib.

For the previous approach, where Storage team was including the header file directly as part of the sources, it was making Storage common to offer the entire JSON API to any customer (like doing PUBLIC linking).  I don't think we want to do that.

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [reply cmake-modules/add_nlohmann_json.cmake:14]: nit: In that case, can we use exact for now, to require precisely the version we need. I don't know if we want to have "silent" upgrades to higher versions because I am not sure if `nlohmnn_json` always guarantees backwards compatibility.

```suggestion
find_package(nlohmann_json 3.8.0 EXACT)
```

**Review by ahsonkhan — APPROVED** [review]
Otherwise, looks good.

**Comment by ahsonkhan** [comment]
Btw, did you observe a noticeable improvement in SDK build times?

File: sdk/core/azure-core/CMakeLists.txt
Comment (vhvb1989) [reply sdk/core/azure-core/CMakeLists.txt:65]: btw, @Jinming-Hu , the link dependencies option keyword is mandatory. There's no default for it.

**Comment by vhvb1989** [comment]
> Btw, did you observe a noticeable improvement in SDK build times?

Not really, looking at the last CI runs, it looks like this one has last less than any other before, but not really too much:
image

Building Storage tests is still taking at least 30min.  I am wondering if we could move this to be only on Storage Live tests and let the core - ci not build storage at all.  just wondering anyway

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:19]: there's possible workaround

```
namspace Details
{
#include <nlohmann/json.hpp>
}
```

File: cmake-modules/add_nlohmann_json.cmake
Comment (Jinming-Hu) [reply cmake-modules/add_nlohmann_json.cmake:20]: Sure, but in this way, you always get a specific version, an old version.

File: sdk/core/azure-core/CMakeLists.txt
Comment (Jinming-Hu) [reply sdk/core/azure-core/CMakeLists.txt:65]: > The difference with PUBLIC is that the library bytes are copied directly to azure-core

I seriously doubt that.

Comment (Jinming-Hu) [reply sdk/core/azure-core/CMakeLists.txt:65]: > the link dependencies option keyword is mandatory. There's no default for it.

I don't think so. https://cmake.org/cmake/help/latest/command/target_link_libraries.html

File: cmake-modules/add_nlohmann_json.cmake
Comment (Jinming-Hu) [reply cmake-modules/add_nlohmann_json.cmake:14]: why not just `find_package(nlohmann_json)`?

How do you know version below 3.8.0 won't work for us?

File: sdk/core/azure-core/CMakeLists.txt
Comment (Jinming-Hu) [reply sdk/core/azure-core/CMakeLists.txt:65]: Really, I don't think it makes too much difference whether we use `INTERFACE` or `PUBLIC` here, or even leave it out.

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [inline cmake-modules/add_nlohmann_json.cmake:5]: FYI, in case it was missed, update this comment before merging.

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [reply cmake-modules/add_nlohmann_json.cmake:14]: > How do you know version below 3.8.0 won't work for us?

See @vhvb1989's comment above. Looks like you are using APIs and features that were added in newer versions, like `contains`:
https://github.com/nlohmann/json/releases/tag/v3.6.0
> introduces a contains function to check if a key is present in an object,
https://github.com/Azure/azure-sdk-for-cpp/blob/71c5cca9d59dd1ab6d61be917090b19a5c4175af/sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp#L1259

> I tried to use an old version (3.2) but Storage was already using 3.8 and got an error because `contains` API was missing, so I updated it.

Comment (ahsonkhan) [reply cmake-modules/add_nlohmann_json.cmake:14]: Not to mention, that's the version of the dependency that's currently checked-in:
https://github.com/Azure/azure-sdk-for-cpp/blob/71c5cca9d59dd1ab6d61be917090b19a5c4175af/sdk/storage/azure-storage-common/inc/azure/storage/common/json.hpp#L4

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [reply cmake-modules/add_nlohmann_json.cmake:20]: Even if we used vcpkg, we'd get a specific pinned version (because we use a specific release of vcpkg), so that aspect remains identical to what @vhvb1989 is doing here. And that's also identical to having a copy of a particular version checked-in.

The version upgrades, like all other dependencies, will need to be intentional.

Comment (Jinming-Hu) [reply cmake-modules/add_nlohmann_json.cmake:20]: But as new versions are released in vcpkg, people can get newer versions, we don't need to do anything.

But since you've written a specific version 3.8.0 in our CMake file https://github.com/Azure/azure-sdk-for-cpp/blob/71d6338d9c91dfc7ef3ee3a0eef335cfafecfa1f/cmake-modules/add_nlohmann_json.cmake#L22. people can only get this specific version, unless we bump up this version, which will increase our maintenance cost.

File: cmake-modules/add_nlohmann_json.cmake
Comment (Jinming-Hu) [reply cmake-modules/add_nlohmann_json.cmake:14]: fine, but shouldn't be 3.6.0 rather than 3.8.0 then?

Also, I believe `EXACT` is not a good idea.

Comment (vhvb1989) [reply cmake-modules/add_nlohmann_json.cmake:10]: ok, will use EXACT + 3.6.0

File: cmake-modules/add_nlohmann_json.cmake
Comment (vhvb1989) [reply cmake-modules/add_nlohmann_json.cmake:16]: Yeah, keep in mind this is not the final decision. Azure Core will still go further about the entire Json story and elaborate about the upgrading strategy and maintenance.

File: sdk/core/azure-core/CMakeLists.txt
Comment (vhvb1989) [reply sdk/core/azure-core/CMakeLists.txt:65]: @Jinming-Hu , did you actually read that link?  
Make sure you select the 3.13 version (https://cmake.org/cmake/help/v3.13/command/target_link_libraries.html)
Then, notice that if we want to omit the KEYWORD (PUBLIC, INTERFACE, PRIVATE), the item we add must be either:
-  library target name
-  A full path to a library file
-  A plain library name
-  A link flag

But what we need is to link an `ALIAS library Target name`. There is even this note in there:
```
Items containing ::, such as Foo::Bar, are assumed to be IMPORTED or ALIAS library target names and will cause an error if no such target exists. See policy CMP0028.
```

Then there is a section describing how to link this by using the Keyword.

Take a look to the next links:
https://stackoverflow.com/questions/26037954/cmake-target-link-libraries-interface-dependencies
https://cmake.org/cmake/help/v3.0/manual/cmake-buildsystem.7.html#transitive-usage-requirements

There is more info about the intention of each KEYWORD. 

Think about it like "Azure CORE is an interface for the JSON lib for others to consume it through it. It is used as a bridge"

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:19]: Yeah, I think this is actually a bug. Headers shouldn't be including this, only cpp files. 
The problem is that this header contains implementation. 
Instead of the Details workaround, Storage team should move the implementation to cpp files and let the header just for the signatures. @Jinming-Hu 

That fix can be done in a different PR since it is not currently the main intention of the actual PR.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp
Comment (vhvb1989) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:16]: Headers files should not include JSON library. Include JSON from each cpp file that uses it.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:19]: https://github.com/Azure/azure-sdk-for-cpp/issues/1009

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [reply cmake-modules/add_nlohmann_json.cmake:10]: > fine, but shouldn't be 3.6.0 rather than 3.8.0 then?

If we are open to changing it, it should be the most recent version.

**Review by ahsonkhan — APPROVED** [review]

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [reply cmake-modules/add_nlohmann_json.cmake:16]: > But as new versions are released in vcpkg, people can get newer versions, we don't need to do anything.

> But since you've written a specific version 3.8.0 in our CMake file

If that is an important scenario to enable, I believe the `CONFIG` setting in `find_package` will match the newer version that is on the user's machine, and we won't be going down this path to download the 3.8.0 version. This path only runs if `find_package` fails.

**Comment by ahsonkhan** [comment]
Other than fixing the merge conflict, looks good as is :)

File: cmake-modules/add_nlohmann_json.cmake
Comment (Jinming-Hu) [reply cmake-modules/add_nlohmann_json.cmake:10]: the most recent version is 3.9.1 https://github.com/nlohmann/json/releases

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:49]: This test change seems unrelated to this PR. Why was it failing and why do we need to sleep here?

File: cmake-modules/add_nlohmann_json.cmake
Comment (ahsonkhan) [inline cmake-modules/add_nlohmann_json.cmake:11]: For future reference, we should not have downgraded from 3.8.0 to 3.6.0, especially in this PR which is replacing the existing checked-in source (which pointed to 3.8.0). Going forward, let's update to newest versions of the library with any bug fixes.

Comment (Jinming-Hu) [reply cmake-modules/add_nlohmann_json.cmake:11]: I still didn't get why we need `EXACT` keyword here.

> Storage requires 3.6.0 version

Actually we require version >= 3.6.0, not exact 3.6.0. Because usually when some library or software adds a new feature in some version, they keep this feature in all the following versions. It's not like they added this feature in 3.6.0 then removed in 3.6.1.

Comment (Jinming-Hu) [reply cmake-modules/add_nlohmann_json.cmake:11]: https://github.com/Azure/azure-sdk-for-cpp/issues/1027

Comment (vhvb1989) [reply cmake-modules/add_nlohmann_json.cmake:11]: EXACT is used to prevent nlohmann/json to add a breaking change in the future that would impact us.

There is a good thing and a bad thing about EXCAT

Good: No breaking changes in the future.
Bad: Azure SDK won't work if the end-user is already using a different version. Unless they remove this EXACT keyword and validate if they can build without issues.

In theory, if we say "we support 3.6 and greater, we would need to test all versions after 3.6 for every release we do to ensure this is true. Otherwise, we are trying to be in a safe zone where checking if the newer version works is something that the end-user needs to try/verify. 

Maybe we can add a build option to disable the EXACT 3rd parties

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:49]: It could have been added in a separate PR, yes. But it was really a super tiny thing!  hahaha

The test started to give some occasionally fails. The test needs to spawn 3 connections by sending 3 requests (one after the other). However, if the first request is completed too fast, then the 3rd request could re-use that connection instead of spawning a new one.  That's why 1 sec sleep will prevent that.
It is just a workaround to make the request to take at least 1 sec to be completed.  
Since this test is not really about if the request is succeeded or not, but to test that a new connection is spawn every time, I didn't invest on making the request adding a payload of something to make it take more time to complete.

File: cmake-modules/add_nlohmann_json.cmake
Comment (Jinming-Hu) [reply cmake-modules/add_nlohmann_json.cmake:11]: Then how about libcurl https://github.com/Azure/azure-sdk-for-cpp/blob/c9c12e27aabe5ce971de33eef22db6d0fc97f0cb/sdk/core/azure-core/CMakeLists.txt#L20? Why don't we try to avoid breaking changes in libcurl?

Comment (ahsonkhan) [reply cmake-modules/add_nlohmann_json.cmake:11]: libcurl provides a compatibility configuration that cmake can use:
https://github.com/curl/curl/blob/f21cc62832a9d5c521d10da9177b075bfed43f2f/CMakeLists.txt#L1514-L1519

See: https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html

It looks like nlohmann json has a similar compatibility declaration, though they don't explicitly use the built-in `SameMajorVersion` cmake literal:
https://github.com/nlohmann/json/blob/fd7a9f600712b2724463e9f7f703878ade676d6e/cmake/nlohmann_jsonConfigVersion.cmake.in#L1

We discussed this a bit today. Though it is useful to do a case-by-case analysis on this topic, based on factors such as pervasiveness, versioning strategy of each dependency or library we use, etc. it is fine to use `config` as the default as a starting point.

Hopefully, either choice here won't have much negative impact to the end user, so it isn't extremely critical, **but we can update this JSON library to use `config`.** This relies on the following assumptions:
1) The library makes an attempt to avoid breaking changes or reduces them as much as possible (looking at its history or stance).
2) When it does make such changes, it follows appropriate semver semantics, and revs the major version.
3) It declares in its cmake package config file that either `AnyNewerVersion` or `SameMajorVersion` (or equivalent) is considered compatible.

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp
Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:49]: Gotcha. That explanation makes sense. Thanks!

nit: whenever we are modifying this file again later, please add this as a comment in the code which indicates why we need the sleep, in that PR.

File: cmake-modules/add_nlohmann_json.cmake
Comment (Jinming-Hu) [reply cmake-modules/add_nlohmann_json.cmake:11]: I don't understand why you guys make so big deal out of it. To me, `find_package(nlohmann_json REQUIRED)` is enough and fits most our customers' needs. But anyway, you're the owner of this code and you decide what to do.

Comment (ahsonkhan) [reply cmake-modules/add_nlohmann_json.cmake:11]: Regardless of code ownership, this conversation has been great since it encourages a deeper understanding of how the various cmake features work, the nuances of our design choices, and the impact on the end user.

**Pull request merged**
